drivers/src/tla2528.cpp

```diff
@@ -0,0 +1,137 @@
+#include <libhal-util/bit.hpp>
+#include <libhal-util/i2c.hpp>
+#include <libhal/error.hpp>
+#include <libhal/units.hpp>
+#include <tla2528.hpp>
+
+namespace sjsu::drivers {
+
+tla2528::tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address)
+  : m_bus(p_i2c)
+{
+  m_i2c_address = p_i2c_address;
+  // TODO: reset command
+}
+
+void tla2528::set_analog_channel(hal::byte p_channel)
+{
+  if (p_channel == m_channel)
+    return;
+  if (p_channel > 7)
+    throw hal::argument_out_of_domain(this);
+
+  std::array<hal::byte, 3> cmd_buffer = { op_codes::single_register_write,
+                                          register_addresses::channel_sel,
+                                          p_channel };
+  hal::write(m_bus, m_i2c_address, cmd_buffer);
+}
+
+void tla2528::set_pin_mode(pin_mode p_mode, hal::byte p_channel)
+{
+  if (p_channel > 7)
+    throw hal::argument_out_of_domain(this);
+  hal::bit_mask channel_mask = hal::bit_mask::from(p_channel);
+  if (p_mode == pin_mode::digital_output_push_pull ||
+      p_mode == pin_mode::digital_output_open_drain) {
+    if (hal::bit_extract(channel_mask, m_object_created) &&
+        !(hal::bit_extract(channel_mask, m_pin_cfg) ||
+          hal::bit_extract(channel_mask, m_gpio_cfg))) {
+      throw hal::resource_unavailable_try_again(this);
+    }
+    hal::bit_modify(m_pin_cfg).set(channel_mask);
+    hal::bit_modify(m_gpio_cfg).set(channel_mask);
+    if (p_mode == pin_mode::digital_output_push_pull) {
+      hal::bit_modify(m_gpo_drive_cfg).set(channel_mask);
+    } else {
+      hal::bit_modify(m_gpo_drive_cfg).clear(channel_mask);
+    }
```

Comment:
What is the purpose of having this as a member variable? It isn't used anywhere else in the code. It could be a temporary variable in the scope of this function. If you are doing this to minimize transactions, I'd recommend performing a read transaction instead. It may take a bit more time, but that eliminates the need to cache information in your class.

Reply:
Same goes for most of the other cached values in the class.

Reply:
as discussed, I removed the pin-type config caches and kept the output value, current channel, and the object created caches.